PEP 730: Revisions and clarifications following community discussion (#3506)
python/peps: peps/pep-0730.rst

* ``"iphonesimulator-arm64"`` for ARM64 simulators
* ``"iphonesimulator-x86_64"`` for x86_64 simulators

``platform``
''''''''''''

Platform be modified to support returning iOS-specific details. Most of the
values returned by the platform module will match those returned by
``os.uname()``, with the exception of:

* ``platform.system()`` - ``"iOS"``, instead of the default ``"Darwin"``

* ``platform.release()`` - the iOS version number, as a string (e.g.,
  ``"16.6.1"``), instead of the Darwin kernel version.

In addition, a ``platform.ios_ver()`` method will be added. This mirrors
``platform.mac_ver()``, which can be used to provide macOS version information.
``ios_ver()`` will return a namedtuple that contains the following:

* ``version`` - the iOS version, as a string (e.g., ``"16.6.1"``).
* ``min_version`` - the minimum supported iOS version, as a string (e.g., ``"12.0"``)

> * ``model`` - the model identifier of the device (e.g., ``iPhone13,2``). On simulators, this
>   will return ``iPhoneSimulator``.

Reviewer: Also a string?
```suggestion
* ``model`` - the model identifier of the device, as a string (e.g., ``iPhone13,2``).
  On simulators, this will return ``"iPhoneSimulator"``.
```

* ``release`` - The Darwin kernel version (e.g., ``"22.6.0"``)

This approach treats the ``os`` module as a "raw" interface to system APIs, and
``platform`` as a higher-level API providing more generally useful values.

``sysconfig``
'''''''''''''

The ``sysconfig`` module will use the minimum iOS version as part of

> ``sysconfig.get_platform()`` identifier (e.g., ``"ios-12.0-iphoneos-arm64"``).

Reviewer 2: ```suggestion
``sysconfig.get_platform()`` (e.g., ``"ios-12.0-iphoneos-arm64"``).
```

== "ios"``.

This finder will convert a Python module name (e.g., ``foo.bar._whiz``) into a

> unique Framework name by using the full Python module as the framework name (i.e.,

Reviewer 2: ```suggestion
unique Framework name by using the full module name as the framework name (i.e.,
```

``foo.bar._whiz.framework``). A framework is a directory; the finder will look
for ``_whiz.dylib`` in that directory.

CI resources
------------

`Anaconda <https://anaconda.com>`__ has offered to provide physical hardware to

> run iOS Buildbots.

Reviewer: ```suggestion
run iOS buildbots.
```
Also below:
```diff
-configured to provide Buildbots or Github Actions runners is limited.
+configured to provide buildbots or GitHub Actions runners is limited.
```

GitHub Actions is able to host iOS simulators on their macOS machines, and the
iOS simulator can be controlled by scripting environments. The free tier
currently only provides x86_64 macOS machines; however ARM64 runners `recently
became available on paid plans <https://github.blog/
2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/>`__.

> However, in order to avoid exhausting macOS runner resources, a Github Actions

Reviewer: ```suggestion
However, in order to avoid exhausting macOS runner resources, a GitHub Actions
```

run for iOS will not be added as part of the standard CI configuration.

Packaging
---------

iOS will not provide a "universal" wheel format. Instead, wheels will be
provided for each ABI-arch combination.

iOS wheels will use tags:

* ``ios_12_0_iphoneos_arm64``
* ``ios_12_0_iphonesimulator_arm64``
* ``ios_12_0_iphonesimulator_x86_64``

In these tags, "12.0" is the minimum supported iOS version. As with macOS, the

> wheel tag will incorporate the minimum iOS version that is selected when CPython

Reviewer 2: ```suggestion
tag will incorporate the minimum iOS version that is selected when the wheel
```

The decision was made to use ``arm64-apple-ios`` and
``arm64-apple-ios-simulator`` because:

1. The ``autoconf`` toolchain already contains support for ``ios`` as a platform
   in ``config.sub``; it's only the simulator that doesn't have a representation.
2. The third part of the host triple is used as ``sys.platform``.
3. When Apple's own tools reference CPU architecture, they use ``arm64``, and the
   GNU tooling usage of the architecture isn't visible outside the build process.
4. When Apple's own tools reference simulator status independent of the OS
   (e.g., in the naming of Swift submodules), they use a ``-simulator`` suffix.
5. While *some* iOS packages will use Rust, *all* iOS packages will use Apple's
   tooling.

"Universal" wheel format
------------------------

macOS currently supports 2 CPU architectures. To aid the end-user development
experience, Python defines a "universal2" wheel format that incorporates both
x86_64 and ARM64 binaries.

It would be conceptually possible to offer an analogous "universal" iOS wheel
format. However, this PEP does not use this approach, for 2 reasons.

Firstly, the experience on macOS, especially in the numerical Python ecosystem,

> as been that universal wheels can be exceedingly difficult to accomodate. While

Reviewer: ```suggestion
has been that universal wheels can be exceedingly difficult to accommodate. While
```

native macOS libraries maintain strong multi-platform support, and Python itself
has been updated, the vast majority of upstream non-Python libraries do not
provide multi-architecture build support. As a result, compiling universal
wheels inevitably requires multiple compilation passes, and complex decisions
over how to distribute header files for different architectures. As a result of
this complexity, many popular projects (including NumPy and Pillow) do not
provide universal wheels at all, instead providing separate ARM64 and x86_64
wheels.

Secondly, historical experience is that iOS would require a much more fluid
"universal" definition. In the last 10 years, there have been *at least* 5
different possible interpretations of "universal" that would apply to iOS,

> including various combinations of armv6, armv7, armv7s, arm64, x86 and and

Reviewer: ```suggestion
including various combinations of armv6, armv7, armv7s, arm64, x86 and
```
